File: trie.py
class DictNode(object):
    def __init__(self):
        self.child = {}
        self.isWord = False

class Trie():
    def __init__(self):
        self.root = DictNode()
    
    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: None
        """
        node = self.root
        for c in word:
            if c not in node.child:
                node.child[c] = DictNode()
            node = node.child[c]
        node.isWord = True
        return
    
    def search(self, word):
        root = self.root
        for c in word:
            if c in root.child:
                root = root.child[c]
            else:
                return False
        return root.isWord
       



troot = Trie()
troot.addWord("bad")
troot.addWord("dad")
troot.addWord("badal")
print troot.search("badass")
print troot.search("badal")
print troot.search("bad")
print troot.root.child

File: trie.cpp


#include<iostream>
#include<vector>

using namespace std;


class Node
{
    public:
        Node *children[26];
        bool isWord;
        Node() {
            memset(children,0,26);
            isWord = false;
        }
};


class Trie
{
    Node *root;
    public:
    Trie() {
        root = new Node();
    }
    void addWord(string);
    bool searchWord(string);
};

void Trie::addWord(string s) {
    Node *node = root;
    cout << "Adding word: " << s;
    for (auto ch : s) {
        int index = ch - 'a';
        cout << index << endl;
        if (node->children[index] == NULL) {
            Node *child = new Node();
            node->children[index] = child;
        }
        
        node = node->children[index];
    }
    node->isWord = true;
    return;
}

bool Trie::searchWord(string s) {
    Node *node = root;
    if (node == NULL)
        return false;

    cout << endl << "searching: " << s;
    for (auto c : s) {
        int index = c - 'a';
        cout << index << endl;
        if (node->children[index] != NULL) {
            node = node->children[index];
        }
        else
            return false;
    }
    return (node->isWord == true);
}




int main() {
    vector<string> strings = {"badal"};
    Trie t = Trie();
    for (auto w: strings) {
        t.addWord(w);
    }
    if ( t.searchWord("badal") ) {
        cout <<"badal present";
    }
    if (! t.searchWord("bada")) {
        cout << "bada nott presentt";
    }
    return 0;
}
